File: algorithms/157.Read_N_Characters_Given_Read4/Read_N_Characters_Given_Read4.java
<<<<<<< HEAD
/* The read4 API is defined in the parent class Reader4.
      int read4(char[] buf); */

public class ReadNCharactersGivenRead4 extends Reader4 {
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    public int read(char[] buf, int n) {
        int index = 0;
        char[] str = new char[4];
        int length = read4(str);
        while (length != 0 && index != n) {
            System.arraycopy( str, 0, buf, index, Math.min(n - index, length));
            index += Math.min(n - index, length);
            length = read4(str);
        }
        return index;
    }
=======
/* The read4 API is defined in the parent class Reader4.
      int read4(char[] buf); */

public class ReadNCharactersGivenRead4 extends Reader4 {
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    public int read(char[] buf, int n) {
        int index = 0;
        char[] str = new char[4];
        int length = read4(str);
        while (length != 0 && index != n) {
            System.arraycopy( str, 0, buf, index, Math.min(n - index, length));
            index += Math.min(n - index, length);
            length = read4(str);
        }
        return index;
    }
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
}

File: algorithms/157.Read_N_Characters_Given_Read4/Read_N_Characters_Given_Read4.py
# The read4 API is already defined for you.
# @param buf, a list of characters
# @return an integer
def read4(buf):
    global file_content
    i = 0
    while i < len(file_content) and i < 4:
        buf[i] = file_content[i]
        i += 1
    
    if len(file_content) > 4:
        file_content = file_content[4:]
    else:
        file_content = ""
    return i
        
class Solution:
    # @param buf, Destination buffer (a list of characters)
    # @param n,   Maximum number of characters to read (an integer)
    # @return     The number of characters read (an integer)
    def read(self, buf, n):
        read_bytes = 0
        eof = False
        buffer = ['' for _ in xrange(4)]
        while not eof and read_bytes < n:
            size = read4(buffer)
            if size < 4:
                eof = True
            bytes = min(n - read_bytes, size)
            for i in xrange(bytes):
                buf[read_bytes + i] = buffer[i]
            read_bytes += bytes
        return read_bytes

if __name__ == "__main__":
    global file_content
    buf = ['' for _ in xrange(100)]
    file_content = "a"
    print buf[:Solution().read(buf, 9)]    
    file_content = "abcdefghijklmnop"
    print buf[:Solution().read(buf, 9)]
